File: topcoder/FloatingMedian.cpp


#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


class FloatingMedian
{

public :

    long long sumOfMedians( long int seed, long int mul, long int add, long int N, long int K )
    {

        long int _tk = seed;
        vector<long int> _temps;
        _temps.push_back( _tk );

        for ( long int q = 0; q < N; q++ )
        {
            _tk = ( _tk * mul + add ) % 65536;
            _temps.push_back( _tk );
        }

        long long _sum = 0;

        vector<long int> _slice;

        for ( long int q = 0; q < N - K + 1; q++ )
        {
            if ( q == 0 )
            {
                for ( long int i = 0; i < K; i++ )
                {
                    _slice.push_back( _temps[i] );
                }
                sort( _slice.begin(), _slice.end() );
            }
            else
            {
                long int _left = _temps[q - 1];
                long int _right = _temps[q + K - 1];
                // cout << vectorIndexOf( _slice, _left ) << endl;
                _slice.erase( _slice.begin() + vectorIndexOf( _slice, _left ) );
                _slice.push_back( _right );
                long int _end = K - 1;
                long int _l = 0;
                long int _r = K - 2;
                long int _mid = ( _l + _r ) / 2;
                if ( _slice[_end] < _slice[_l] )
                {
                    _slice.insert( _slice.begin(), _slice[_end] );
                    _slice.pop_back();
                }
                else if ( _slice[_end] <= _slice[_r] )
                {
                    while ( true )
                    {
                        if ( _slice[_end] > _slice[_mid] )
                        {
                            _l = _mid;
                        }
                        else
                        {
                            _r = _mid;
                        }
                        _mid = ( _l + _r ) / 2;
                        if ( _l >= _mid )
                        {
                            break;
                        }
                    }
                    _slice.insert( _slice.begin() + _r, _slice[_end] );
                    _slice.pop_back();
                }
            }
            long int _median = _slice[ ( K - 1 ) / 2 ];
            // printVector( _slice );
            _sum += _median;
            // cout << "med: " << _median << endl;
        }

        return _sum;
    } 

    long int vectorIndexOf( const vector<long int>& vect, long int val )
    {
        long int _indx = -1;

        for ( long int q = 0; q < vect.size(); q++ )
        {
            if ( vect[q] == val )
            {
                _indx = q;
                break;
            }
        }

        return _indx;
    }

    void printVector( const vector<long int>& vect )
    {
        cout << "[ ";
        for ( long int q = 0; q < vect.size(); q++ )
        {
            cout << vect[q] << " ";
        }
        cout << "]" << endl;
    }

};






int main()
{

    FloatingMedian _prob;

    cout << "-> " << _prob.sumOfMedians( 3, 1, 1, 10, 3 ) << endl;
    cout << "-> " << _prob.sumOfMedians( 10, 0, 13, 5, 2 ) << endl;
    cout << "-> " << _prob.sumOfMedians( 4123, 2341, 1231, 7, 3 ) << endl;
    cout << "-> " << _prob.sumOfMedians( 47, 5621, 1, 125000, 1700 ) << endl;
    cout << "-> " << _prob.sumOfMedians( 32321, 46543, 32552, 17, 17 ) << endl;


}

File: topcoder/FloatingMedian.py


class FloatingMedian:


    def sumOfMedians( self, seed, mul, add, N, K ) :
        _k = 0
        _tk = seed
        # generate the temperatures
        _temps = [ _tk ]
        for q in range( 1, N ) :
            _tk = ( _tk * mul + add ) % 65536
            _temps.append( _tk )
        # print _temps
        # take the median of each slice of size k
        _sum = 0
        _slice = None
        for i in range( 0, N - K + 1 ) :
            if i == 0 : ## just in the first case sort
                _sub = _temps[i:i + K]
                _sub.sort()
                _slice = _sub
                # find the median in slice [i - i+k-1]
            else :
                _left = _temps[i - 1]
                _right = _temps[i + K - 1]
                ## sort by removing the left most and adding the rightmost
                #_slice.remove( _left )
                #_slice.append( _right )
                q = K - 1
                while ( q > 0 ) :
                    if ( _slice[q] < _slice[q - 1] ) :
                        _tmp = _slice[q]
                        _slice[q] = _slice[q - 1]
                        _slice[q - 1] = _tmp
                    q -= 1

            _median = _slice[ ( len( _slice ) - 1 ) / 2 ]
            _sum += _median
            # print _median
        return _sum


_prob = FloatingMedian()

print _prob.sumOfMedians( 3, 1, 1, 10, 3 )
print _prob.sumOfMedians( 10, 0, 13, 5, 2 )
print _prob.sumOfMedians( 4123, 2341, 1231, 7, 3 )
print _prob.sumOfMedians( 47, 5621, 1, 125000, 1700 )
print _prob.sumOfMedians( 32321, 46543, 32552, 17, 17 )








        
